os/drivers/input/touchscreen.c

```diff
@@ -0,0 +1,358 @@
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <sys/types.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+#include <poll.h>
+#include <assert.h>
+#include <errno.h>
+#include <debug.h>
+#include <tinyara/fs/fs.h>
+#include <tinyara/fs/ioctl.h>
+#include <tinyara/input/touchscreen.h>
+
+/****************************************************************************
+ * Pre-Processor Definitions
+ ****************************************************************************/
+
+#define EVENT_PACKET_SIZE 16
+
+#ifndef CONFIG_TOUCHL_NPOLLWAITERS
+#define CONFIG_TOUCH_NPOLLWAITERS 2
+#endif
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int touch_open(FAR struct file *filep);
+static int touch_close(FAR struct file *filep);
+static ssize_t touch_read(FAR struct file *filep, FAR char *buffer, size_t buflen);
+static ssize_t touch_write(FAR struct file *filep, FAR const char *buffer, size_t buflen);
+static int touch_poll(FAR struct file *filep, FAR struct pollfd *fds, bool setup);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static const struct file_operations g_touchdev_fileops = {
+	touch_open,		/* open */
+	touch_close,	/* close */
+	touch_read,		/* read */
+	touch_write,	/* write */
+	NULL,			/* seek */
+	NULL,			/*ioctl */
+	touch_poll,
+};
+
+/************************************************************************************
+ * Name: touch_semtake
+ ************************************************************************************/
+
+static inline int touch_semtake(FAR sem_t *sem, bool errout)
+{
+	/* Take a count from the semaphore, possibly waiting */
+	if (sem_wait(sem) < 0) {
+		/* EINTR is the only error that we expect */
+		int errcode = get_errno();
+		DEBUGASSERT(errcode == EINTR);
+		if (errout) {
+			return errcode;
+		}
+	}
```

Comment:
```suggestion
	while (sem_wait(sem) != OK) {
		ASSERT(get_errno() == EINTR);
		if (errout) {
			return -EINTR;
		}
	}
```

sem_wait can be woken up by other signal. this case should be call sem_wait again

Reply:
done